pkg/mutation/cache.go

```diff
@@ -0,0 +1,193 @@
+package mutation
+
+import (
+	"fmt"
+	"sort"
+	"sync"
+
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+// Cache represent a mutation cache containting all the mutation resources
+type Cache struct {
+	mutators []Mutator
+	sync.RWMutex
+	scheme *runtime.Scheme
+}
+
+// NewCache initializes an empty cache
+func NewCache(scheme *runtime.Scheme) *Cache {
+	return &Cache{
+		mutators: make([]Mutator, 0),
+		scheme:   scheme,
+	}
+}
+
+// Insert inserts the mutation resource into the cache
+func (c *Cache) Insert(mutator Mutator) error {
+	c.Lock()
+	defer c.Unlock()
+	var lastErr error
+	i := sort.Search(len(c.mutators), func(i int) bool {
+		res, err := c.greater(c.mutators[i], mutator)
+		if err != nil {
+			lastErr = err
+		}
+		return res
+	})
+	if lastErr != nil {
+		return errors.Wrap(lastErr, "Error while inserting in cache")
+	}
+
+	if i == len(c.mutators) { // Adding to the bottom of the list
+		c.mutators = append(c.mutators, mutator)
+		return nil
+	}
+	updating, err := c.same(c.mutators[i], mutator)
+	if err != nil {
+		return errors.Wrap(err, "Error while checking for equality in cache")
+	}
+
+	if updating {
+		c.mutators[i] = mutator
+		return nil
+	}
+
+	c.mutators = append(c.mutators, nil)
+	copy(c.mutators[i+1:], c.mutators[i:])
+	c.mutators[i] = mutator
+	return nil
+}
+
+// Remove removes the mutation object from the cache
+func (c *Cache) Remove(mutator Mutator) error {
+	c.Lock()
+	defer c.Unlock()
+	var lastErr error
+	i := sort.Search(len(c.mutators), func(i int) bool {
+		res, err := c.same(c.mutators[i], mutator)
```

Comment:
Why are we not stopping at the first error ?  could there be a state where objectes are inserted illegality ?  (not saying this should be removed though)

Reply:
Because the signature of the function accepted to sort.Search is fixed, and search can't be stopped.
So I use a closure to track that I had at least one error making the whole function to fail.
`sort.Search` is not inserting anything, it just returns the index where the item should go.